src/async_io/async_io.ml

```diff
@@ -0,0 +1,303 @@
+        in
+        let should_interrupt =
+          match Table.find table t.fd with
+          | None -> false
+          | Some q ->
+            let new_q = Queue.create () in
+            Queue.iter q ~f:(fun (Task t' as task) ->
+                if Task_id.equal t.id t'.id then Scheduler.cancel_job_started ()
+                else Queue.push new_q task);
+            if Queue.is_empty new_q then (
+              Table.remove table t.fd;
+              Queue.is_empty q)
+            else (
+              Table.add_exn table t.fd new_q;
+              Queue.length new_q <> Queue.length q)
+        in
+        if should_interrupt then interrupt t.select;
+        Fiber.Ivar.fill t.ivar (Error `Cancelled)
+    in
+    Mutex.unlock t.select.mutex
+end
+
+let drain_until_ready queue acc =
+  match Queue.pop queue with
+  | None -> acc
+  | Some (Task task) ->
+    let result = try Ok (task.job ()) with exn -> Error (`Exn exn) in
+    Fiber.Fill (task.ivar, result) :: acc
+
+let make_fills fds pipe_fd waiters init =
+  List.fold_left fds ~init:(false, init) ~f:(fun (pipe, acc) fd ->
+      if Fd.equal fd pipe_fd then (true, acc)
+      else
+        let acc =
+          match Table.find waiters fd with
+          | None -> acc
+          | Some w ->
+            let acc = drain_until_ready w acc in
+            if Queue.is_empty w then Table.remove waiters fd;
+            acc
+        in
+        (pipe, acc))
+
+let rec drain_pipe pipe buf =
+  (* a single read should be enough, but just in case *)
+  match Unix.read pipe buf 0 (Bytes.length buf) with
+  | 0 -> ()
+  | _ -> drain_pipe pipe buf
+  | exception Unix.Unix_error (Unix.EAGAIN, _, _) -> ()
+
+let rec drain_cancel q acc =
+  match Queue.pop q with
+  | None -> acc
+  | Some (Task task) ->
+    drain_cancel q (Fiber.Fill (task.ivar, Error `Cancelled) :: acc)
+
+let maybe_cancel table fd acc =
+  match Table.find table fd with
+  | None -> acc
+  | Some q ->
+    Table.remove table fd;
+    drain_cancel q acc
+
+let rec select_loop t =
+  (match t.to_close with
+  | [] -> ()
+  | to_close -> (
+    let fills =
+      List.fold_left to_close ~init:[] ~f:(fun acc fd ->
+          Unix.close fd;
+          let acc = maybe_cancel t.readers fd acc in
+          maybe_cancel t.writers fd acc)
+    in
+    t.to_close <- [];
+    match fills with
+    | [] -> ()
+    | _ :: _ ->
+      let module Scheduler = (val t.scheduler) in
+      Scheduler.fill_jobs fills));
+  match t.running with
+  | false ->
+    Unix.close t.pipe_write;
+    Unix.close t.pipe_read
+  | true ->
+    let readers, writers, ex =
+      let read = t.pipe_read :: Table.keys t.readers in
+      let write = Table.keys t.writers in
+      Mutex.unlock t.mutex;
+      (* At this point, if any [ready] acquires the lock, they need to check if
+         [read] or [write] contain their fd. If it doesn't, the write
+         [t.pipe_write] will interrupt this select *)
+      Unix.select read write [] (-1.0)
```

Comment:
According to its manual: https://www.man7.org/linux/man-pages/man2/select.2.html
```
WARNING: select() can monitor only file descriptors numbers that
       are less than FD_SETSIZE (1024)—an unreasonably low limit for
       many modern applications—and this limitation will not change.
       All modern applications should instead use [poll(2)](https://www.man7.org/linux/man-pages/man2/poll.2.html) or [epoll(7)](https://www.man7.org/linux/man-pages/man7/epoll.7.html),
       which do not suffer this limitation.
```
Is that a concern here?

Reply:
Not too much of a concern. The thread based implementation had similar scalability limitations.

Although eventually we'll want to graduate away from select as well.